app/cluster/state/etcd.go

```diff
@@ -0,0 +1,289 @@
+}
+
+// Ping ensures the connection to etcd is alive
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+
+	_, err := manager.Client.Cluster.MemberList(context.Background())
+	if err != nil {
+		return fmt.Errorf("ping: get cluster member list: %w", err)
+	}
+
+	return nil
+}
+
+func (manager *ETCDManager) prepareMode(raw []byte) servermode.ChangeEvent {
+	var req modeRequestValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return servermode.ChangeEventError(fmt.Errorf("unmarshal mode request: %w", err))
+	}
+
+	mode := req.Mode
+	if !mode.Valid() {
+		return servermode.ChangeEventError(fmt.Errorf("invalid mode: %s", mode))
+	}
+
+	return servermode.NewChangeEvent(
+		mode,
+		func() error {
+			ackValue, err := json.MarshalToString(modeAckValue{
+				Status: mode,
+			})
+			if err != nil {
+				return fmt.Errorf("marshal ack value: %w", err)
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			if err != nil {
+				return fmt.Errorf("put value to ack key %q: %w", req.AckKey, err)
+			}
+			return err
+		})
+}
+
+func errChModeRequest(err error) <-chan servermode.ChangeEvent {
+	ch := make(chan servermode.ChangeEvent, 1)
+	ch <- servermode.ChangeEventError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) ServerMode(ctx context.Context) <-chan servermode.ChangeEvent {
+	if err := manager.init(); err != nil {
+		return errChModeRequest(err)
+	}
+
+	modeRequestKey := fmt.Sprintf(modeRequestKeyPattern, manager.Config.Namespace, manager.Config.ServerIndex)
+
+	resultChan := make(chan servermode.ChangeEvent, 1)
+	resp, err := manager.Client.Get(ctx, modeRequestKey)
+	if err != nil {
+		return errChModeRequest(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errChModeRequest(errors.New("no workspace found"))
+	}
+
+	resultChan <- manager.prepareMode(resp.Kvs[0].Value)
+
+	etcdWatchChan := manager.Client.Watch(ctx, modeRequestKey, clientv3.WithRev(resp.Header.Revision+1))
+	go func() {
+		for watchResp := range etcdWatchChan {
+			if watchResp.Err() != nil {
+				resultChan <- servermode.ChangeEventError(watchResp.Err())
+				continue
+			}
+
+			for _, event := range watchResp.Events {
+				switch event.Type {
+				case mvccpb.PUT:
+					resultChan <- manager.prepareMode(event.Kv.Value)
+				default:
+					resultChan <- servermode.ChangeEventError(fmt.Errorf("unknown event type %q", event.Type))
+				}
+			}
+		}
+		close(resultChan)
+	}()
+
+	return resultChan
+}
+
+func errChWorkspacesRequest(err error) <-chan workspace.ChangeEvent {
+	ch := make(chan workspace.ChangeEvent, 1)
+	ch <- workspace.ChangeEventError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) prepareWorkspace(raw []byte) workspace.ChangeEvent {
+	var req workspacesRequestsValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return workspace.ChangeEventError(err)
+	}
+
+	return workspace.NewWorkspacesRequest(
+		strings.Split(req.Workspaces, ","),
```

Comment:
Is the splitting just to let the reader know that we are serving multiple workspaces?

Reply:
Yes, I think is better to move this as an array for two reasons:
1. Readability: _I expect many ids, thus an array/slice_
2. Decouple transport layer with business logic: the fact that we are using commas to separate the workspace ids is specific to implementation, we might as well use a JSON array, or `;` - separated fields. To prevent the core logic of the system to care about this is better to do the parsing here.